### PR TITLE
[IMP] hr_timesheet, project_timesheet_holidays, sale_timesheet: using create_multi for batch creation

### PR DESCRIPTION
**Pull request by parthchoksi-github**
**Creation of records will be done in batches.**

- Used model_create_multi
- Reworked on related functions(for eg: _timesheet_postprocess) so as to make it work in batches.

This optimization is done in order to speed up the creation.

Task Link- https://www.odoo.com/web?debug#id=1971460&action=327&model=project.task&view_type=form&menu_id=4720

Pad Link- https://pad.odoo.com/p/r.3817b8906a9ee8d5ebbf45b55b89a845

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 1]
```diff
@@ -63,21 +63,23 @@ def _compute_department_id(self):
...
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            # compute employee only for timesheet lines, makes no sense for other lines
+            if not vals.get('employee_id') and vals.get('project_id'):
+                if vals.get('user_id'):
+                    ts_user_id = vals['user_id']
+                else:
+                    ts_user_id = self._default_user()
+                vals['employee_id'] = self.env['hr.employee'].search([('user_id', '=', ts_user_id)], limit=1).id
```
Comment (jem-odoo) [inline addons/hr_timesheet/models/hr_timesheet.py]: Isn't there a way to extract this search outside of the loop ? Or maybe use the `employee_id` field on the user ?

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 2]
```diff
@@ -63,21 +63,23 @@ def _compute_department_id(self):
...
+            vals.update(self._timesheet_preprocess(vals))
+
+        lines = super(AccountAnalyticLine, self).create(vals_list)
+        for line, values in zip(lines, vals_list):
```
Comment (jem-odoo) [inline addons/hr_timesheet/models/hr_timesheet.py]: :+1:

File: addons/project_timesheet_holidays/models/hr_holidays.py [hunk 3]
```diff
@@ -73,7 +74,7 @@ def _validate_leave_request(self):
                 fields.Datetime.from_string(holiday.date_to),
             )
             for index, (day_date, work_hours_count) in enumerate(work_hours_data):
-                self.env['account.analytic.line'].sudo().create({
+                analytic_line_vals.append({
```
Comment (jem-odoo) [inline addons/project_timesheet_holidays/models/hr_holidays.py]: 1/ we are talking of timesheet here, so avoid `analytic_line_` prefix ;)
2/ you probably need to rebase this, as code change in 13.0

File: addons/account/models/account_move.py [hunk 4]
```diff
@@ -3398,18 +3398,20 @@ def create_analytic_lines(self):
...
-                    self.env['account.analytic.line'].create(vals_line)
+                    analytic_line_vals.append(vals_line)
             if obj_line.analytic_account_id:
                 lines_to_create_analytic_entries |= obj_line
 
         # create analytic entries in batch
         if lines_to_create_analytic_entries:
-            values_list = lines_to_create_analytic_entries._prepare_analytic_line()
-            self.env['account.analytic.line'].create(values_list)
+            analytic_line_vals += lines_to_create_analytic_entries._prepare_analytic_line()
```
Comment (jem-odoo) [inline addons/account/models/account_move.py]: I would prefer `.append` rather than `+` to stay consistent.

Comment (pch-odoo) [reply addons/account/models/account_move.py]: @jem-odoo Actually the `_prepare_analytic_line()` is returning **list** of values. 
Hence += is used. To use `.append`, we will have to make changes in return values of `_prepare_analytic_line()`
So I preferred going through with +=

File: addons/project_timesheet_holidays/models/hr_holidays.py (hunk 3 above)
Comment (pch-odoo) [reply addons/project_timesheet_holidays/models/hr_holidays.py]: Done :+1:

File: addons/hr_timesheet/models/hr_timesheet.py (hunk 1 above)
Comment (pch-odoo) [reply addons/hr_timesheet/models/hr_timesheet.py]: Done :+1:

**Comment by tivisse** [comment]
@robodoo r+

**Pull request closed without merging**